position_in_partition.hh

```diff
@@ -676,3 +676,41 @@ bool position_range::overlaps(const schema& s, position_in_partition_view start,
     position_in_partition::less_compare less(s);
     return !less(end, _start) && less(start, _end);
 }
+
+// Track position_in_partittion transitions
+// and validate monotonicity.
+class position_in_partition_tracker {
```

Comment:
I suggest to make it a `Filter` (as defined by flat_mutation_reader.hh).

This way you only need to implement:

```
bool operator()(const mutation_fragment& mf);
bool operator()(const dht::decorated_key& dk);
```

And work in terms of `mf.position()`. No need to even know what are the possible fragment types (see https://github.com/tgrabiec/scylla/commit/fa53994fb04697b3e1e60060fe62122fc577b106).

In sstables.cc, call `consume_in_thread()` variant which accepts a `Filter`.

Reply:
Isn't this a lot more heavy-weight (also compares clustering keys)? Or is this something we want to do eventually anyway? In that case it is indeed better to just compare positions.

Even if so, I think this is stretching the filter concept a bit far.

Reply:
Hmm, sounds intriguing...

Reply:
@denesb You can compare just the `region()` of a position. You still don't need to care about fragment types.

What's the problem with using a filter? It would be a filter, one which doesn't filter anything, but still, it's a filter. The advantage is that we reuse the existing infrastructure. We avoid the risk of missing a callback, plus the patch is smaller.

Reply:
Right. It would indeed be much simpler.

Reply:
Using a Filter class in v2 as @tgrabiec suggested.